src/targets/browser/browserLauncher.ts

```diff
@@ -140,25 +147,80 @@ export class BrowserLauncher implements ILauncher {
+    telemetryReporter: TelemetryReporter,
+  ): Promise<number> {
+    const promisedPort = getDeferred<number>();
+
+    if (!params.runtimeExecutable) {
+      // runtimeExecutable is required for web view debugging.
+      promisedPort.resolve(params.port);
+      return promisedPort.promise;
     }
+
+    const exeName = params.runtimeExecutable.split(/\\|\//).pop();
+    const pipeName = `VSCode_${randomBytes(12).toString('base64')}`;
+    // This is a known pipe name scheme described in the web view documentation
+    // https://docs.microsoft.com/microsoft-edge/hosting/webview2/reference/webview2.idl
+    const serverName = `\\\\.\\pipe\\WebView2\\Debugger\\${exeName}\\${pipeName}`;
+
+    const server = createServer(stream => {
+      stream.on('data', async data => {
+        const info: IWebViewConnectionInfo = JSON.parse(data.toString());
+
+        // devtoolsActivePort will always start with the port number
+        // and look something like '92202\n ...'
+        const dtString = info.devtoolsActivePort || '';
+        const dtPort = parseInt(dtString.split('\n').shift() || '');
+        const port = params.port || dtPort || params.port;
+
+        if (!this._mainTarget && filter(info)) {
+          promisedPort.resolve(port);
+        }
+
+        // All web views started under our debugger are waiting to to be resumed.
+        const wsURL = `ws://${params.address}:${port}/devtools/${info.type}/${info.id}`;
+        const ws = await WebSocketTransport.create(wsURL, NeverCancelled);
+        const connection = new CdpConnection(ws, telemetryReporter);
+        await connection.rootSession().Runtime.runIfWaitingForDebugger({});
+        connection.close();
+      });
+    });
+    server.on('close', () => promisedPort.resolve(params.port));
+    server.listen(serverName);
+
+    // We must set a user data directory so the DevToolsActivePort file will be written.
+    // See: https://crrev.com//21e1940/content/public/browser/devtools_agent_host.h#99
+    params.userDataDir =
+      params.userDataDir || path.join(tmpdir(), `vscode-js-debug-userdatadir_${params.port}`);
+
+    // Web views are indirectly configured for debugging with environment variables.
+    // See the WebView2 documentation for more details.
+    params.env = params.env || {};
+    params.env['WEBVIEW2_USER_DATA_FOLDER'] = params.userDataDir.toString();
     params.env['WEBVIEW2_ADDITIONAL_BROWSER_ARGUMENTS'] = `--remote-debugging-port=${params.port}`;
     params.env['WEBVIEW2_WAIT_FOR_SCRIPT_DEBUGGER'] = 'true';
+    params.env['WEBVIEW2_PIPE_FOR_SCRIPT_DEBUGGER'] = pipeName;
+
+    return promisedPort.promise;
   }
 
   async prepareLaunch(
     params: IChromeLaunchConfiguration,
     { dap, targetOrigin, cancellationToken, telemetryReporter }: ILaunchContext,
     clientCapabilities: IDapInitializeParamsWithExtensions,
   ): Promise<BrowserTarget | string> {
+    const filter = createTargetFilterForConfig(params);
+    const promisedPort =
```

Comment:
It looks like this has some compiler errors since if `params.useWebView` evaluates to false `promisedPort` will be a boolean, but if it's true `promisedPort` will be a `Promise<number>`.

In any case, does the webview logic belong here, since it's very specific to Edge? I'm thinking it would be cleaner and easier to maintain by maybe creating an `EdgeLauncher` class which extends from `BrowserLauncher` which can take care of Edge specific cases and leave `BrowserLauncher` more agnostic, and also avoid the Edge webview logic causing any unintended side effects when launching Chrome.

Reply:
>
> It looks like this has some compiler errors since if params.useWebView evaluates to false promisedPort will be a boolean, but if it's true promisedPort will be a Promise<number>.

This is actually OK in TypeScript, the `promised` port becomes a union type of `false | Promise<number>` which must be checked for truthiness before we can use it as a promise, [check out this example on the TS playground][1].

> In any case, does the webview logic belong here, since it's very specific to Edge?

I agree that a better home for the web view logic will eventually be another class, ideally it's own even separate from Edge, that said your suggestion of an `EdgeLauncher` would be a move in the right direction.

My intention with this change is to get the feature working with the existing classes as a first step and benefit to users of the extension and I created issue #266 to follow-up track splitting out Edge from Chrome debugger targets, addressing the concerns your mentioned.

[1]: https://www.typescriptlang.org/play/#code/CYUwxgNghgTiAEYD2A7AzgF3hkmBc8ARkkhCFCgNwBQ1y6WAlgAoxIC2jaCAvNrlgBkg+ChAB3eKw5cQAHkwxGKAOYA+ABQx4PNfBgaA5KTIALQwEoLNahqhoAnijDwNFnXoDe1eL-hksFTgQDGUVAkUwmz94IJAQsJ14KHEoRiZpTm5KGPgAejzRJCwkAGsfP0YAM1cWNiyQd29cvziE1SSUtIz62RyC+DKK3wBfWhHrWiA

Reply:
> ...and I created issue #266 to follow-up track splitting out Edge from Chrome debugger targets

Fair enough

> This is actually OK in TypeScript, the promised port becomes a union type of false | Promise<number>

While true, I'd still argue that it's confusing code. Also, I get compile errors on this line when I try to build your PR branch locally, but CI seems to build fine 🤷‍♂ (the error I get is actually on the filter param, not the promisedPort type)
![image](https://user-images.githubusercontent.com/38218905/73498532-6cdd5a00-4372-11ea-895e-3b7fa86cf5ec.png)

Reply:
> While true, I'd still argue that it's confusing code. 

Thanks for the feedback on this, I coded it for brevity but agree that overlying on short-circuits reduces readability. I will rewrite it with an `if` statement.

> I get compile errors on this line when I try to build your PR branch locally, but CI seems to build fine 🤷‍♂

This is because my change relies on #283 which I haven't merged in from master on my branch, sorry about that! CI passes because it is running on the merge of my branch + master.

I will rebase my change with your feedback and master so it will build locally (thanks for testing!).

Reply:
Done!